### PR TITLE
fix: use --track-unscheduled-pods to select unscheduled pods in Daemonset sharding

### PR DESCRIPTION
**Pull request by CatherineF-dev**
**What this PR does / why we need it**:

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kube-state-metrics/issues/2353

Tested using minikube

### CODE REVIEW

**Comment by mrueg** [comment]
Will this supersede https://github.com/kubernetes/kube-state-metrics/pull/2373 ?

**Comment by CatherineF-dev** [comment]
> Will this supersede https://github.com/kubernetes/kube-state-metrics/pull/2373 ?

Yes.

File: pkg/options/types.go
Comment (mrueg) [inline pkg/options/types.go:148]: Can you add a comment here, why we're handling this case specially?

Comment (CatherineF-dev) [reply pkg/options/types.go:149]: Added.

**Comment by mrueg** [comment]
/hold 
for others to review.
/lgtm

**Review by logicalhan — COMMENTED** [review]
tests?

**Comment by diranged** [comment]
@logicalhan Can we just skip that for now? We've been waiting for 6 weeks to https://github.com/kubernetes/kube-state-metrics/issues/2374 resolved, and we're having _daily_ pain points with KSM right now in our environment. While I appreciate adding tests, we're literally removing code and making it simpler here.. can we bypass this and just get a fix in ASAP?

**Comment by logicalhan** [comment]
> @logicalhan Can we just skip that for now? We've been waiting for 6 weeks to https://github.com/kubernetes/kube-state-metrics/issues/2374 resolved, and we're having _daily_ pain points with KSM right now in our environment. While I appreciate adding tests, we're literally removing code and making it simpler here.. can we bypass this and just get a fix in ASAP?

How do you know the fix works without tests?

**Comment by diranged** [comment]
> > @logicalhan Can we just skip that for now? We've been waiting for 6 weeks to #2374 resolved, and we're having _daily_ pain points with KSM right now in our environment. While I appreciate adding tests, we're literally removing code and making it simpler here.. can we bypass this and just get a fix in ASAP?
> 
> How do you know the fix works without tests?

I think this is one of those vanity test situations... I don't see it worth blocking on this just to get a code coverage line. If you have a test proposal, would you put it up? Otherwise, can we get this feature fixed since it's been broken?

**Comment by logicalhan** [comment]
> I think this is one of those vanity test situations... I don't see it worth blocking on this just to get a code coverage line. If you have a test proposal, would you put it up? Otherwise, can we get this feature fixed since it's been broken?

This bug exists because we didn't have test coverage what the heck are you talking about. The answer is no.

**Comment by logicalhan** [comment]
If you want it so badly @diranged then you should write some tests. We accept pulls.

**Comment by diranged** [comment]
> If you want it so badly @diranged then you should write some tests. We accept pulls.

Given that I'm not deeply familiar with the code - that seems like a poor idea. I wish you would reconsider here how much you want to stand on the 'needs tests' statement for such a simple PR.

**Comment by mrueg** [comment]
@CatherineF-dev do you have plans to add tests to this PR to get it into a mergable state?

**Comment by CatherineF-dev** [comment]
Okay! Work in Progress

File: tests/e2e.sh
Comment (LaikaN57) [inline tests/e2e.sh:116]:
```suggestion
    cat ${KUBE_STATE_METRICS_LOG_DIR}/daemonset-scraped-metrics ${KUBE_STATE_METRICS_LOG_DIR}/deployment-scraped-metrics ${KUBE_STATE_METRICS_LOG_DIR}/deployment-scraped-no-node-metrics >> ${KUBE_STATE_METRICS_LOG_DIR}/all-pod-metrics
```

Fixes:
```
In ./tests/e2e.sh line 114:
    cat ${KUBE_STATE_METRICS_LOG_DIR}/daemonset-scraped-metrics >> ${KUBE_STATE_METRICS_LOG_DIR}/all-pod-metrics
    ^-- SC21[29](https://github.com/kubernetes/kube-state-metrics/actions/runs/9390258474/job/25859669104?pr=2388#step:5:30) (style): Consider using { cmd1; cmd2; } >> file instead of individual redirects.

For more information:
  https://www.shellcheck.net/wiki/SC2129 -- Consider using { cmd1; cmd2; } >>...In ./tests/e2e.sh line 114:
    cat ${KUBE_STATE_METRICS_LOG_DIR}/daemonset-scraped-metrics >> ${KUBE_STATE_METRICS_LOG_DIR}/all-pod-metrics
    ^-- SC21[29](https://github.com/kubernetes/kube-state-metrics/actions/runs/9390258474/job/25859669104?pr=2388#step:5:30) (style): Consider using { cmd1; cmd2; } >> file instead of individual redirects.

For more information:
  https://www.shellcheck.net/wiki/SC2129 -- Consider using { cmd1; cmd2; } >>...
```

https://github.com/kubernetes/kube-state-metrics/actions/runs/9390258474/job/25859669104?pr=2388#step:5:28

File: pkg/options/types.go
Comment (LaikaN57) [inline pkg/options/types.go:143]: To make the test pass and keep most of the logic, I think we can just revert this line.

```suggestion
	if nil == n || len(*n) == 0 {
```

I was trying to look at the usage of this function in [pkg/app/server.go#L208](https://github.com/kubernetes/kube-state-metrics/blob/7995d5fd23bcff7ae24ab6849f7c393d262fb025/pkg/app/server.go#L208).

... However, I still don't fully understand what the intended behavior is when `--node` is not provided.

Comment (LaikaN57) [reply pkg/options/types.go:143]: After reading https://github.com/kubernetes/kube-state-metrics/pull/2359 I believe this is the correct fix.

File: data.yaml
Comment (mrueg) [inline data.yaml:4]: Please do not bump a version in a bugfix.

File: tests/e2e/testdata/pods.yaml
Comment (mrueg) [inline tests/e2e/testdata/pods.yaml:30]: where is this image from? will the cluster attempt to download it?

File: data.yaml
Comment (CatherineF-dev) [reply data.yaml:4]: Ok, I was testing whether e2e is using latest KSM or not.

File: tests/e2e/testdata/pods.yaml
Comment (LaikaN57) [inline tests/e2e/testdata/pods.yaml:29]: typo?

```
$ docker run -it --rm -p 8080:8080 registry.k8s.io/e2e-test-images/agnhost:2.39 netexec --http-port=8080q
Error: invalid argument "8080q" for "--http-port" flag: strconv.ParseInt: parsing "8080q": invalid syntax
```

Comment (LaikaN57) [inline tests/e2e/testdata/pods.yaml:29]:
```suggestion
        - --http-port=8080
```

File: tests/e2e/testdata/pods.yaml
Comment (LaikaN57) [reply tests/e2e/testdata/pods.yaml:30]: > where is this image from?

The question is a bit unclear but, [kubernetes/kubernetes (/test/images/agnhost)](https://github.com/kubernetes/kubernetes/tree/master/test/images/agnhost#image) seams to be the source.

> will the cluster attempt to download it?

Again, the question is a bit unclear but, no end-user clusters will not download this; it is just used for e2e testing AFAIK.

Comment (mrueg) [reply tests/e2e/testdata/pods.yaml:30]: Apologies for not being precise, the reason I was asking mainly is:
* Will this image be downloaded to a node? (= can I run e2e tests in an airgapped environment, or when developing on a plane?)
* Is the image source stable (= will CI integration tests break)?

**Comment by LaikaN57** [comment]
@CatherineF-dev Is there anything I can do to help here? We had good momentum last week and I am looking to help continue that!

How do you feel about my comments for fixing the unit tests? (ie. adding back `|| len(*n) == 0` check)

I naively feel that fixing that will help out with e2e tests but I'm sure you have way more context working on this and other related code / architecture.

**Comment by CatherineF-dev** [comment]
I was looking at the E2E test failure https://github.com/kubernetes/kube-state-metrics/actions/runs/9417050145/job/25941445490?pr=2388

File: docs/developer/cli-arguments.md
Comment (mrueg) [inline docs/developer/cli-arguments.md:49]: Suggestion: Move this closer to the `--node` argument and rename from no_node to unscheduled_pods or similar?

```suggestion
      --node-sharding-unscheduled-pods-only                     This configuration is used in conjunction with node configuration. When this configuration is true, node configuration is empty and the metric of unscheduled pods is scraped from the Kubernetes API. This is experimental.
```

File: README.md
Comment (mrueg) [inline README.md:296]: If you don't want to prefix it by --node (no strong opinion here), can we use something like: 
```suggestion
        - --enable-unscheduled-pods-fetching
```
Using gerund instead of infinitive style (not to mix both)

File: docs/developer/cli-arguments.md
Comment (mrueg) [inline docs/developer/cli-arguments.md:49]: Should the description make it more clear that this is only useful if you're doing sharding?

File: pkg/options/types_test.go
Comment (mrueg) [inline pkg/options/types_test.go:189]: Can we add a few more tests here? What happens if I set --node and the --fetch-unscheduled-pods argument at the same time? Should it warn about this?

**Comment by CatherineF-dev** [comment]
@LaikaN57 fixed all tests now.

File: pkg/options/types.go
Comment (LaikaN57) [inline pkg/options/types.go:112]:
```suggestion
func (n *NodeType) GetNodeFieldSelector(fetchUnscheduledPods bool) string {
	if fetchUnscheduledPods {
```

nit: Function argument name was a bit unclear. It was describing the effect of the argument not the cause/condition it represented.

File: pkg/options/types.go
Comment (LaikaN57) [inline pkg/options/types.go:108]: Because this is no longer a complex type should we just remove this type alias so we don't have to cast in other places?

**Comment by LaikaN57** [comment]
Thanks @CatherineF-dev! I am OOO until Monday, but I did a quick review of the new code. Just 2 small nits but nothing blocking. I can review more fully on Monday if others do not approve by then. 🤞 I relly appreciate all the work you put into this!

File: docs/developer/cli-arguments.md
Comment (mrueg) [inline docs/developer/cli-arguments.md:49]:
```suggestion
      --enable-unscheduled-pods-fetching           This configuration is used in conjunction with node configuration. When this configuration is true, node configuration is empty and the metric of unscheduled pods is fetched from the Kubernetes API Server. This is experimental.
```

File: jsonnet/kube-state-metrics/kube-state-metrics.libsonnet
Comment (mrueg) [inline jsonnet/kube-state-metrics/kube-state-metrics.libsonnet:384]:
```suggestion
      local shardksmname = ksm.name + "-unscheduled-pods-fetching";
```

**Comment by CatherineF-dev** [comment]
@mrueg need a review again. Rebased to HEAD.

**Comment by LaikaN57** [comment]
> tests?

@logicalhan Tests now implemented thanks to all of @CatherineF-dev's hard work 🙏 Looks like we are looking for review again now. Thanks!

File: pkg/options/types.go
Comment (rexagod) [reply pkg/options/types.go:108]: This seems necessary since there seem to be methods defined on it below.

File: pkg/options/types.go
Comment (rexagod) [inline pkg/options/types.go:112]: Let's keep the [`flag.Flag`](https://pkg.go.dev/flag#Value) behavior.

File: tests/e2e/testdata/pods.yaml
Comment (rexagod) [reply tests/e2e/testdata/pods.yaml]: I believe the image is stable ([scheduled releases](https://github.com/kubernetes/kubernetes/commits/master/test/images/agnhost/VERSION)), and safe for us to use in the given context. LMK if I missed anything though.

File: pkg/options/options.go
Comment (dgrisonnet) [inline pkg/options/options.go:139]: Thinking about it again, would it really make sense for any user to want to opt-out of collection metrics from unscheduled pods?

As a user, I would expect to collect the same metrics whether I am using uuid sharding, node sharding or even an unsharded installation of ksm. So based on that I am wondering if we should really make this a flag and instead just enable that by default when node sharding is enabled.

Comment (LaikaN57) [reply pkg/options/options.go:139]: I think the issue is that as soon as you node shard, the current code will only select node-bound pods. Therefore you would run an additional deployment of KSM with this option enabled specifically. I don't think there is an easy way in the current codebase for us to filter pods on the current node and unbound pods at the same time. But I will leave this to the experts, this was just my current understanding.

Comment (dgrisonnet) [reply pkg/options/options.go:139]: Ah indeed you are correct, I forgot we had to differentiate it from the other shards.

Comment (mrueg) [reply pkg/options/options.go:139]: So we should still set it to true by default and disable when sharding?

File: pkg/options/types.go
Comment (dgrisonnet) [inline pkg/options/types.go:118]: It feels odd to handle this here since fetchUnscheduledPods is not related to the node object. I think we should take care of this scenario in the caller `RunKubeStateMetrics`

File: pkg/options/types.go
Comment (rexagod) [reply pkg/options/types.go:112]: It seems https://github.com/kubernetes/kube-state-metrics/pull/2388#discussion_r1655039540 was resolved without a fix? Can you PTAL?

File: docs/developer/cli-arguments.md
Comment (dgrisonnet) [inline docs/developer/cli-arguments.md:49]: format

**Review by rexagod — CHANGES_REQUESTED** [review]
PTAL https://github.com/kubernetes/kube-state-metrics/pull/2388#discussion_r1655039540.

File: pkg/options/types.go
Comment (LaikaN57) [reply pkg/options/types.go:112]: > Let's keep the [`flag.Flag`](https://pkg.go.dev/flag#Value) behavior.

Did you mean [`flag.Value`](https://pkg.go.dev/flag#Value)?

Comment (rexagod) [reply pkg/options/types.go:112]: Right, sorry, I meant that.

**Comment by mrueg** [comment]
Heads up: This is currently listed for the v2.13 milestone. In order to get a release out soon, I'll drop all open PRs by Wednesday from that milestone.

**Comment by CatherineF-dev** [comment]
@rexagod have fixed the change request.  Thx! Need a lgtm

**Comment by CatherineF-dev** [comment]
@rexagod fixed.

**Comment by dgrisonnet** [comment]
LGTM

File: pkg/options/options.go
Comment (mrueg) [inline pkg/options/options.go:160]: Is the type conversion needed?

**Comment by jeffmccune** [comment]
@CatherineF-dev @mrueg @dgrisonnet FYI, I ran into some very easy to overlook syntax errors in this CL that eagle-eyed `jsonnet` complained loudly about when trying to customize kube-prometheus.

 Please take a look at https://github.com/kubernetes/kube-state-metrics/pull/2454 for a quick fix.

**Pull request merged**
